### PR TITLE
Durable users - WIP / for discussion

### PR DESCRIPTION
**Pull request by devth**
Opening this up as a place to discuss how we want to model users. 

## Goals of this discussion 

1. Provide a consistent and durable model of users to build on top of
   - Karma would gain a foreign key to users 
   - GraphQL resources like history and karma would be able to link to the user entity
   - We would be able to reliably build [user prefs](https://github.com/yetibot/yetibot/issues/427) on top of it
1. Figure out how and if Yetibot should support multiple adapters. Users are the primary entity that we need to consider, but all db state should be evaluated.

### Desired state

### CODE REVIEW

**Comment by devth** [comment]
From @jcorrado via Slack:

> I'm still of the "all of nothing, do it cleanly or don't do it at all" mindset (edited) 
but, I understand practical middle-ground  (in theory)
for me the simple thought experiment goes: were we to simplify to one adapter, what would that simplified, clarified architecture open up?
and is that worth more than the what would be lost

**Comment by devth** [comment]
> were we to simplify to one adapter, what would that simplified, clarified architecture open up?

Corollary: What if we just design everything as if there were only one adapter (while still capturing provenance)? Rationale:

1. I'm probably only person running multi adapter, so every operator of Yetibot aside from myself would still have clean single-adapter state and
1. we wouldn't have to invest any time into stitching user accounts and making everything truly multi adapter

This is essentially what I've been doing all along and it hasn't hindered me. The remaining TODOs for Users would still be:

1. add provenance to all state (good idea, regardless)
1. model users, including provenance

I think of it as a stepping stone to the ideal "all or nothing" state.

This means both sets of users show up in karma. Observers, Status, and Aliases are global. I think all of these are fine. I've always tried to let concrete use cases drive the design of Yetibot, and I don't think I have a strong enough multi adapter case (Yetibot public runs on Slack and IRC but no one uses it on IRC currently) to help drive its design while at the same time having a practical need for multi adapter.

In the future if/when YB supports Gitter or other platforms it will likely listen on all, and I think we want most state to be global in that case. Same with Clojure community chats: if/when we get Yetibot into Clojurians, #clojure IRC, clojure/general on Gitter, etc. When that happens I think I'll either have motivation to go all in or the clarity and confidence to make the call go single adapter.

**Comment by jcorrado** [comment]
Actively mulling this over...

**Comment by devth** [comment]
tldr: agree with all or nothing but want to defer the decision / work.

**Comment by devth** [comment]
We discussed this today starting at https://yetibot.slack.com/archives/C66AJ2EFL/p1554821184926700

## Summary

1. Our goal is "true" multi-adapter
2. We seem to be in agreement on delivering fully-baked, user-proof features.  Quality UI.
3. Even with #2 in mind, we're practical about how we build, and are willing to cut some internal corners to deliver those features, when there's a compelling opportunity.  sic: Clean features but fine to 'fix up' internals later
4. Clojurians Slack is something we'd like to service, ideally soon to coincide with Clojure/North.  We're OK levering its curtailed requirements ("Just Slack for now"), so long as features are clean.  We can fix up later.  This #2 and #3 _in action_.

## Short term

1. Require Slack-encoded users in the karma command. This lets us get around the fact that we don't have durable users modeled yet.
1. Disable karma in IRC unless we can easily support it (@jcorrado will investigate)
1. Consider removing or disabling users command until it's properly modeled

## User modeling

1. create or update durable users when the bot observes their existence: startup, channel join/leave events, polling if necessary, or lazy create-on-demand e.g. when a user runs a command or mentions another user in a command (like karma)
1. check existence of user in the db when trying to attribute karma, and handle adapter-specific user encodings.

## Mid term

1. add provenance (chat adapter and channel) to all db state to give us flexibility on partitioning or uniting things later on
1. begin storing users "as seen by the bot" in database for both irc and slack (might require some discussion/planning on its own)
1. don't worry about leaking state across adapters for now
1. consider in the back of our minds that someday we might have the notion of a single identity spread across multiple adapters (possibly an `identities` table that joins 1 or more rows from `users`)

**Comment by devth** [comment]
Related: https://github.com/yetibot/yetibot/issues/179

File: src/yetibot/core/commands/alias.clj [hunk 1]
```diff
@@ -22,7 +22,8 @@
           ;; evaluation of piped expressions (i.e. %s instead of $s)
           expr (binding [*subst-prefix* method-like-replacement-prefix]
                  (str command/config-prefix (pseudo-format-n cmd args)))
-          results (record-and-run-raw expr user yetibot-user
+          resolved-users nil ;; TODO
```
Comment (devth) [inline src/yetibot/core/commands/alias.clj]: I'm not sure aliases would ever mention users. Either way, we can defer this to a later PR if we want.

File: src/yetibot/core/commands/observe.clj [hunk 2]
```diff
@@ -104,7 +104,8 @@
                                   ;; piping behavior
                                   rendered-cmd))
                       [{:keys [error? timeout? result] :as expr-result}]
-                      (record-and-run-raw expr user yetibot-user)]
+                      resolved-users nil ;; TODO
```
Comment (devth) [inline src/yetibot/core/commands/observe.clj]: Can defer.

File: src/yetibot/core/adapters/slack.clj [hunk 3]
```diff
@@ -234,6 +234,9 @@
           yetibot? (= yetibot-uid (:user event))
           user-model (assoc (users/get-user cs (:user event))
                             :yetibot? yetibot?)
+          ;; TODO - call (a/resolve-users event) to ensure users are resolved,
+          ;; then pass the resulting map along to `handle-raw` as
+          ;; `:resolved-users` below in addition to `:body`.
```
Comment (devth) [inline src/yetibot/core/adapters/slack.clj]: This is a potential starting place. 

1. Add `a/resolve-users` function
2. call it via `(a/resolve-users adapter event)`
3. in `a/resolve-users` log the `event` and start looking at how we can pull users out of it and turn them into canonical persistent records
